addons/editor-dark-mode/text_color.js

```diff
@@ -0,0 +1,110 @@
+  const { r, g, b } = parseHex(hex);
+  threshold = threshold !== undefined ? threshold : 170;
+  if (r * 0.299 + g * 0.587 + b * 0.114 > threshold) {
+    // https://stackoverflow.com/a/3943023
+    setCSSVar(varName, black !== undefined ? black : "#575e75");
+  } else {
+    setCSSVar(varName, white !== undefined ? white : "#ffffff");
+  }
+}
+
+function transparentVariant(varName, hex, opacity) {
+  const { r, g, b } = parseHex(hex);
+  setCSSVar(varName, `rgba(${r}, ${g}, ${b}, ${opacity})`);
+}
+
+function lightDarkVariant(varName, hex, threshold, cr, cg, cb, lr, lg, lb) {
+  const { r, g, b } = parseHex(hex);
+  if (lr === undefined) lr = cr;
+  if (lg === undefined) lg = cg;
+  if (lb === undefined) lb = cb;
+  textColor(
+    varName,
+    hex,
+    `rgb(${cr * r}, ${cg * g}, ${cb * b})`,
+    `rgb(${(1 - lr) * 255 + lr * r}, ${(1 - lg) * 255 + lg * g}, ${(1 - lb) * 255 + lb * b})`,
+    threshold
+  );
+}
+
+function testAll(settings) {
+  for (let setting of [
+    "page",
+    "primary",
+    "menuBar",
+    "selector",
+    "selector2",
+    "selectorSelection",
+    "accent",
+    "input",
+    "categoryMenu",
+    "palette",
+  ]) {
+    textColor(`${setting}-text`, settings.get(setting));
+  }
+  textColor("page-tabHoverFilter", settings.get("page"), "grayscale(100%)", "brightness(0) invert(1)");
+  textColor("primary-filter", settings.get("primary"), "brightness(0.4)", "none");
+  textColor("primary-filter2", settings.get("primary"), "none", "brightness(0) invert(1)");
+  textColor("secondary-filter", settings.get("secondary"), "brightness(0.4)", "none");
+  textColor("menuBar-transparentText", settings.get("menuBar"), "rgba(87, 94, 177, 0.25)", "rgba(255, 255, 255, 0.25)");
+  textColor("menuBar-dimText", settings.get("menuBar"), "rgba(87, 94, 177, 0.75)", "rgba(255, 255, 255, 0.75)");
+  textColor("menuBar-filter", settings.get("menuBar"), "brightness(0.4)", "none");
+  textColor("menuBar-border", settings.get("menuBar"), "rgba(0, 0, 0, 0.15)", "rgba(255, 255, 255, 0.15)", 60);
+  textColor("tab-text", settings.get("tab"), "rgba(87, 94, 117, 0.75)", "rgba(255, 255, 255, 0.75)");
+  textColor("selector2-filter", settings.get("selector2"), "none", "brightness(0) invert(1)");
+  textColor("accent-filter", settings.get("accent"), "none", "brightness(0) invert(1)");
+  textColor("accent-desaturateFilter", settings.get("accent"), "saturate(0)", "brightness(0) invert(1)");
+  textColor("input-transparentText", settings.get("input"), "rgba(87, 94, 117, 0.6)", "rgba(255, 255, 255, 0.4)");
+  textColor("input-filter", settings.get("input"), "none", "brightness(0) invert(1)");
+  textColor("input-codeZoomFilter", settings.get("input"), "none", "invert(1) hue-rotate(180deg)");
+  textColor(
+    "categoryMenu-selection",
+    settings.get("categoryMenu"),
+    "rgba(87, 124, 155, 0.13)",
+    "rgba(255, 255, 255, 0.05)"
+  );
+  transparentVariant("primary-transparent35", settings.get("primary"), "0.35");
```

Comment:
I'm not sure this code (`transparentVariant` being passed `settings.get`) is working? I noticed this when I looked at the colors around a selected sprite:

![a selected sprite](https://user-images.githubusercontent.com/9948030/111079020-e54f9d80-84d6-11eb-87bb-e83236756d11.png)

The proper contrast around the (thin) border of the element and the (thick) box shadow isn't there, even though the colors are set appropriately (`--editorDarkMode-primary` and `--editorDarkMode-primary-transparent35`).

The RGB value of `primary` is, as expected, `#47566b`. But the RGB of `primary-transparent35` is `#4d97ff`. (This is a much brighter color - gauging by the preview, possibly the default 3.0 primary? - so when it's reduced to 35% alpha, it looks pretty much the same as the correctly overridden primary color against a dark background.)

(PS, I'm testing with the Dark Editor theme. I'm the author of it, and I love seeing how it's been worked into this PR! It's a lot more refined than my original version :) )

Reply:
I think it needs to be `addon.settings.get`

Reply:
Another place this shows up is in the tag menu bar (also `primary-transparent35`):

![Two stacked headers in the costume modal. Top one is primary, bottom is broken transparent color (slightly saturated blue).](https://user-images.githubusercontent.com/9948030/111080316-b3413a00-84dc-11eb-8d5e-9a9661d8b4e5.png)

And here's an example of the "action menu" colors not being overridden (`primary-variant` & `primary-transparent35`):

!["Choose a Sprite" action menu. The slide-out portion is a medium-bright blue. The ring around the main button is the usual incorrect transparent color.](https://user-images.githubusercontent.com/9948030/111080290-8c830380-84dc-11eb-85a1-a16cc175def5.png)

Reply:
The shadow of the main action menu button not being overridden is a bug, I'll fix it soon. Other instances of primary color variations work for me - does it work correctly when you reload?

> (PS, I'm testing with the Dark Editor theme. I'm the author of it, and I love seeing how it's been worked into this PR! It's a lot more refined than my original version :) )

I think the most noticeable changes are the use of `#111111` as the border color for all elements to avoid adding another setting and the darker primary color which I changed based on your comment [here](https://github.com/LLK/scratch-gui/issues/59#:~:text=I'd%20love%20to%20hear%20more%20as%20to%20why%20you%20made%20%28the%20add,similar%20to%20what%20I%20did%20for%20the%20menu%20bar.).

Reply:
I found the code which was broken on my end :) https://github.com/ScratchAddons/ScratchAddons/pull/1805#discussion_r593962470

I think it was still working for you because you had the "secondary" setting internally stored, but from a fresh install (i.e. mine) that setting had been removed and so wasn't present, hence the error.

Reply:
Fixed.